builders/interact/ethereum-api/libraries/web3py.md

```diff
@@ -0,0 +1,264 @@
+
+```bash
+python3 deploy.py
+```
+
+If successful, the contract's address will be displayed in the terminal.
+
+![Deploy Contract Web3py](/images/builders/interact/ethereum-api/web3py/web3py-3.png)
+
+### Read Contract Data (Call Methods) {: #read-contract-data }
+
+Call methods are the type of interaction that don't modify the contract's storage (change variables), meaning no transaction needs to be sent. They simply read various storage variables of the deployed contract.
+
+To get started, you can create a file and name it `get.py`:
+
+```bash
+touch get.py
+```
+
+Then you can take the following steps to create the script:
+
+1. Add imports, including Web3.py and the ABI of the `Incrementer.sol` contract
+2. [Set up the Web3 provider](#setting-up-the-web3py-provider)
+3. Define the `account_from`, including the `private_key`. The private key is required to sign the transaction. **Note: This is for example purposes only. Never store your private keys in a Python file**
+4. Create a contract instance using the `web3.eth.contract` function and passing in the ABI and address of the deployed contract
+5. Using the contract instance, you can then call the `number` function
+
+```python
+--8<-- 'code/ethereum-api/web3py/get.py'
+```
+
+To run the script, you can enter the following command in your terminal:
+
+```bash
+python3 get.py
+```
+
+If successful, the value will be displayed in the terminal.
+
+![Read from Contract Web3py](/images/builders/interact/ethereum-api/web3py/web3py-4.png)
+
+### Interact with Contract (Send Methods) {: #interact-with-contract }
+
+Send methods are the type of interaction that modify the contract's storage (change variables), meaning a transaction needs to be signed and sent. In this section, you'll create two scripts: one to increment and one to reset the incrementer. To get started, you can create a file for each script and name them `increment.py` and `reset.py`:
+
+```bash
+touch increment.py reset.py
+```
+
+Open the `increment.py` file and take the following steps to create the script:
+
+1. Add imports, including Web3.py and the ABI of the `Incrementer.sol` contract
+2. [Set up the Web3 provider](#setting-up-the-web3py-provider)
+3. Define the `account_from`, including the `private_key`, the `contract_address` of the deployed contract, and the `value` to increment by. The private key is required to sign the transaction. **Note: This is for example purposes only. Never store your private keys in a Python file**
+4. Create a contract instance using the `web3.eth.contract` function and passing in the ABI and address of the deployed contract
+5. Build the increment transaction using the contract instance and passing in the value to increment by. You'll then use the `build_transaction` function to pass in the transaction information including the `from` address and the `nonce` for the sender. To get the `nonce` you can use the `web3.eth.get_transaction_count` function
```

Comment:
This step fails (at least on my computer) when calling the method "web3.eth.get_transaction_count", with exception web3.exceptions.InvalidAddress: ('web3.py only accepts checksum addresses blah...')
To solve this, converting the address to its checksum version works:
web3.eth.get_transaction_count( Web3.to_checksum_address(account_from['address']) )

Reply:
Looks like the 'from' parameter from the tx also requires a checksum address. the complete script could be changed to something like:
# 5. Build increment tx
checksum_address = Web3.to_checksum_address(account_from['address'])
increment_tx = Incrementer.functions.increment(value).build_transaction(
    {
        'from': checksum_address,
        'nonce': web3.eth.get_transaction_count(checksum_address),
    }
)

Reply:
I don't hit this issue but I can add it so that it is compatible with more versions/OS. Updated both snippets for Increment and Reset files